lib/cancel-stuck-txs.js

```diff
@@ -0,0 +1,171 @@
+      gasFeeCap: string;
+      gasPremium: string;
+      method: string;
+      methodNumber: number;
+      evmMethod: string;
+      createTimestamp: number;
+    }[];
+  }}
+  */
+  const { messages } = /** @type {any} */(await res.json())
+  return messages
+}
+
+/**
+ * @returns {Promise<{
+  "cid": string;
+  "height": number;
+  "timestamp": number;
+  "gasLimit": number;
+  "gasFeeCap": string;
+  "gasPremium": string;
+  "method": string;
+  "methodNumber": number;
+  "receipt": {
+      "exitCode": number;
+      "return": string;
+      "gasUsed": number;
+    },
+    "size": number;
+    "error": string;
+    "baseFee": string;
+    "fee": {
+      "baseFeeBurn": string;
+      "overEstimationBurn": string;
+      "minerPenalty": string;
+      "minerTip": string;
+      "refund": string;
+    },
+  }>}
+*/
+async function getRecentSendMessage () {
+  let res = await fetch('https://filfox.info/api/v1/message/list?method=Send')
+  if (!res.ok) {
+    throw new Error(`Filfox request failed with ${res.status}: ${(await res.text()).trimEnd()}`)
+  }
+  const body = /** @type {any} */(await res.json())
+  assert(body.messages.length > 0, '/message/list returned an empty list')
+  const sendMsg = body.messages.find(m => m.method === 'Send')
+  assert(!!sendMsg, 'No Send message found in the recent committed messages')
+  const cid = sendMsg.cid
+
+  res = await fetch(`https://filfox.info/api/v1/message/${cid}`)
+  if (!res.ok) {
+    throw new Error(`Filfox request failed with ${res.status}: ${(await res.text()).trimEnd()}`)
+  }
+
+  return /** @type {any} */(await res.json())
+}
+
+const cancelStuckTxs = async ({ walletDelegatedAddress, signer }) => {
+  console.log('Checking for stuck transactions...')
+
+  const messages = await getMessagesInMempool(walletDelegatedAddress)
+  const txsToCancel = messages.filter(m => {
+    return new Date(m.createTimestamp * 1000) <
+      new Date(Date.now() - (2 * ROUND_LENGTH_MS))
+  })
+  if (txsToCancel.length === 0) {
+    console.log('No transactions to cancel')
+    return false
+  }
+
+  console.log('Transactions to cancel:')
+  for (const tx of txsToCancel) {
+    console.log(
+      '-',
+      tx.cid,
+      `(age ${ms(Date.now() - (tx.createTimestamp * 1000))})`
+    )
+  }
+
+  const recentSendMessage = await getRecentSendMessage()
+  console.log('Calculating gas fees from the recent Send message %s (created at %s)',
+    recentSendMessage.cid,
+    new Date(recentSendMessage.timestamp * 1000).toISOString()
+  )
+
+  const gasUsed = recentSendMessage.receipt.gasUsed
+  const gasFeeCap = Number(recentSendMessage.gasFeeCap)
+
+  await Promise.all(txsToCancel.map(async tx => {
+    const oldGasPremium = Number(tx.gasPremium)
+    const nonce = tx.nonce
+
+    console.log(`Replacing ${tx.cid}...`)
+    try {
+      const replacementTx = await signer.sendTransaction({
+        to: signer.address,
+        value: 0,
+        nonce,
+        gasLimit: Math.ceil(gasUsed * 1.1),
```

Comment:
Should we extract the magic constant `1.1` into a named constant defined at the top of the file to make it more clear what knobs we can turn to change the behaviour of the auto-cancellation process?

Also, could you please add a code comment to explain to future readers how to choose this value?

Reply:
I didn't come up with this number, I don't know the reason why it's 1.1 tbh

Reply:
Tbh, 1.1 doesn't make sense to me, wrt Riba's comment, which states:

> you also must LOWER your gas-limit to match the new gas requirements of the message you are sending:
if you do not do so, the network will penalize you for lying ( you said this will cost X gas but
used 1/10th of X to execute: no good )

Is increasing the gas limit value actually lowering the gas limit?

Reply:
Hmm, this is confusing, isn't it? 

Here is my understanding: We need to lower the gas limit from the limit calculated for the expensive smart contract call transaction to the limit that's reasonable for a cheap SEND transaction.

To do so, it should be enough to set `gasLimit=recentGasUsed` (set the limit to the real amount of gas used by a recent SEND transaction).

Then there are two questions:
1. Why do we need to increase this value?
2. Why 10% and not a different ratio?

I _think_ that the answer is to account for a possible rise in the base fee since the last SEND transaction and the 10% value is a heuristic - a tradeoff between having a high confidence (but paying too much) vs paying just enough with the risk of the replacement transaction getting stuck again.

WDYT?

Reply:
Hmm yeah that could be, it makes sense like this I think